Cython/Build/Tests/TestInline.py

```diff
@@ -110,3 +114,88 @@ def test_numpy(self):
         a[0,0] = 10
         self.assertEqual(safe_type(a), 'numpy.ndarray[numpy.float64_t, ndim=2]')
         self.assertEqual(inline("return a[0,0]", a=a, **self.test_kwds), 10.0)
+
+
+class TestCompileDecorator(CythonInlineTest):
+    def test_no_kwargs(self):
+        old_cache_dir = os.environ.get('CYTHON_CACHE_DIR', '')
+        os.environ['CYTHON_CACHE_DIR'] = self.test_kwds['lib_dir']
+        try:
+            # We need to make sure that compilation works when there are no
+            # arguments to the decorator
+            @cython_compile
+            def test_fn(x):
+                return x + 1
+        finally:
+            os.environ['CYTHON_CACHE_DIR'] = old_cache_dir
+
+        self.assertEqual(test_fn(1), 2)
+
+    def test_kwargs(self):
+
+        @cython_compile(**self.test_kwds)
+        def test_fn(x=":"):
+            return x + 1
+
+        self.assertEqual(test_fn(1), 2)
+
+    def test_no_decorator(self):
+
+        def test_fn(x):
+            return x + 1
+
+        # The decorator should work as a normal function, too
+        test_fn = cython_compile(**self.test_kwds)(test_fn)
+
+        self.assertEqual(test_fn(1), 2)
+
+        # It should also work when there's another decorator
+        @boundscheck(False)
+        def test_fn(x):
+            return x + 1
+
+        test_fn = cython_compile(**self.test_kwds)(test_fn)
+
+        self.assertEqual(test_fn(1), 2)
+
+    @unittest.skipIf(not has_numpy, "NumPy is not available")
+    def test_numpy(self):
+
+        import numpy as np
+
+        # Make sure the aliased module dependency is handled properly
+        @cython_compile(**self.test_kwds)
+        def test_fn(x):
+            return np.add(x, 1)
+
+        self.assertEqual(test_fn(1), 2)
+
+    @unittest.skipIf(sys.version_info < (3, 6), "Requires Py3.6+")
+    def test_type_hints(self):
+        fd, tmp_code_file = tempfile.mkstemp(suffix='.py', text=True)
+        with os.fdopen(fd, 'w') as f:
+            tmp_code = textwrap.dedent("""
+            def test_fn(x: int, y: float) -> int:
+                b: int = 1
+                return x + int(y) + b
+            """)
+            f.write(tmp_code)
+
+        locals = {}
+        exec(compile(tmp_code, tmp_code_file, 'exec'), None, locals)
+
+        assert 'test_fn' in locals
+        test_fn = locals['test_fn']
+
+        test_fn = cython_compile(**self.test_kwds)(test_fn)
+
+        self.assertEqual(test_fn(1, 2), 4)
+
+    def test_decorator_with_decorator(self):
+
+        @cython_compile(**self.test_kwds)
+        @boundscheck(False)
+        def test_fn(x):
+            return x[0]
```

Comment:
It would really help if the decorator was actually useful and testable. Maybe try the `@always_allow_keywords(False)` decorator with a single argument (`METH_O`) function, and then pass a keyword argument into it. That should fail if the decorator is correctly applied.

Reply:
I was looking for such a decorator when I made this test, so I'll try that out.